File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* solve(ListNode* head, int k, int &len) {
        // base case 
        if(len < k || head == NULL) {
            return head;
        }
        
        // reverse logic
        ListNode* prev = NULL;
        ListNode* curr = head;
        
        int temp = k;
        while(temp > 0) {
            ListNode* tempNode = curr -> next;
            curr -> next = prev;
            prev = curr;
            curr = tempNode;
            
            temp--;
        }
        
        int updatedLen = len - k;
        
        ListNode* remHead = solve(curr, k, updatedLen);
        
        head -> next = remHead;
        
        return prev;
    }
    
    ListNode* reverseKGroup(ListNode* head, int k) {
        // using recursive appraoch 
        if(head == NULL || head -> next == NULL) return head;
        
        // calculate length of the linked list
        ListNode* curr = head;
        int len = 0;
        
        while(curr != NULL) {
            len++;
            curr = curr -> next;
        }
        
        ListNode* newHead = solve(head, k, len);
        
        return newHead;
    }
};

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {    
    dHead, curr, i := head, head, 0
    
    if k == 1 {
        return head
    }

    for curr != nil {
        i++
        if i % k == 0 {
            // for first reversal 
            if dHead == head {
                reverse(dHead, curr)
                head = curr
                curr = dHead.Next
            } else {
                reverse(dHead.Next, curr)
                tmp := curr
                tmp2 := dHead.Next
                curr = dHead.Next.Next
                dHead.Next = tmp
                dHead = tmp2
            }
        } else {
            curr = curr.Next
        }
    }
    
    return head
}

func reverse(start *ListNode, end *ListNode) {
    // next node after group
    node := end.Next
    dummy := &ListNode{Next: start}
    prev := dummy
    
    for start != node {
        tmp := start.Next
        start.Next = prev
        prev = start
        start = tmp
        
    }
    
    dummy.Next.Next = node
    dummy.Next = nil
}